### PR TITLE
CDAP-11997 Clear FileSystem.CACHE when closing an explore operation.

### PR DESCRIPTION
**Pull request by anwar6953**
[CDAP-11997](https://issues.cask.co/browse/CDAP-11997)

FileSystem.CACHE adds a new FileSystem entry to the cache, for each new UserGroupInformation that is accessing a FileSystem via FileSystem.get(...). Because of how we do impersonation, there are a log of different UGI objects being created, and so this cache can grow continuously.

See [HIVE-9234](https://issues.apache.org/jira/browse/HIVE-9234) for a similar fix.

https://builds.cask.co/browse/CDAP-RUT1161-2

### CODE REVIEW

**Comment by anwar6953** [comment]
Tested with a LoggingThread, that logs the size of the cache each second (source below). (The thread can be started in BaseHiveExploreService#startUp).

Creating a stream increases the size of the cache by 2.
Executing a 'show tables' query increases the size of the cache by 2.
Executing a 'select *' query against a PartitionedFileSet increases the size of the cache by 2, and previewing/downloading the results increases it by 1.
Without the fix, the size of the cache would only go up, but with the fix, it comes back down. to a steady number (2-3).

```java
private static final class LoggingThread extends Thread {
    @Override
    public void run() {
      LOG.info("Launching logging thread.");
      while (true) {
        try {
          logCacheSize();
          TimeUnit.SECONDS.sleep(1);
        } catch (InterruptedException e) {
          break;
        } catch (Exception e) {
          LOG.error("Got exception.", e);
        }
      }
    }

    private void logCacheSize() throws NoSuchFieldException, IllegalAccessException {
      Field cacheField = FileSystem.class.getDeclaredField("CACHE");
      cacheField.setAccessible(true);
      Object cache = cacheField.get(null);

      Map map = get(cache, "map");
      Set toAutoClose = get(cache, "toAutoClose");
      LOG.info("map.size(): {}, toAutoClose.size(): {}", map.size(), toAutoClose.size());
    }

    private <T> T get(Object o, String fieldName) throws IllegalAccessException, NoSuchFieldException {
      Class<?> clz = o.getClass();
      Field field = clz.getDeclaredField(fieldName);
      field.setAccessible(true);
      return (T) field.get(o);
    }
  }
```

**Comment by chtyim** [comment]
LGTM

File: cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/ActiveOperationRemovalHandler.java [hunk 1]
```diff
@@ -91,6 +94,16 @@ public Void call() throws Exception {
         });
       } catch (Exception e) {
         LOG.error("Failed to impersonate while closing handle {}", handle);
+      } finally {
+        try {
+          // we don't want to close FileSystems for the login user
+          if (!UserGroupInformation.getLoginUser().equals(opInfo.getUGI())) {
+            FileSystem.closeAllForUGI(opInfo.getUGI());
```
Comment (poornachandra) [inline cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/ActiveOperationRemovalHandler.java]: Would be good to add a comment on why this is needed. Also can you add a note to https://issues.cask.co/browse/CDAP-12011 to move this cleanup to happen when the UGI is invalidated from the cache.

Comment (anwar6953) [reply cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/ActiveOperationRemovalHandler.java]: done

**Review by poornachandra — APPROVED** [review]

**Pull request merged**